# EXERCISE 1
## Part 1.1.

In [5]:
import numpy as np
import pandas as pd
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV

In [6]:
df = pd.read_csv('trspam.csv', on_bad_lines='skip')
df.head(10)

Text Classification
0  Sayın Yetkili ,\n  Gelişen ve değişen günümüz ...           spam
1  Sayın Yetkili,\n \n28 Kasım 2010 tarihli KPSS ...           spam
2  Sayın Yetkili ,\n  Gelişen ve değişen günümüz ...           spam
3  T.C.\nURLA KAYMAKAMLIĞI\nURLA HAKAN ÇEKEN ANAD...           spam
4  Sayın Yetkili ,Hızla büyüyen ve gelişmekte ola...           spam
5  Sayın Yetkili ,\n\nHızla büyüyen ve gelişmekte...           spam
6  Bizler ÖSYM’nin 10--11 Temmuz 2010 tarihinde y...           spam
7  Sayın Yetkili ,Hızla büyüyen ve gelişmekte ola...           spam
8  Sayın Yetkili ,Hızla büyüyen ve gelişmekte ola...           spam
9  Sayın Yetklili ,En fazla mezun veren bölümlerd...           spam

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 701 entries, 0 to 700
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Text            700 non-null    object
 1   Classification  701 non-null    object
dtypes: object(2)
memory usage: 11.1+ KB


In [95]:
pd.unique(df['Classification'])

array(['spam', 'ham'], dtype=object)

In [97]:
df = df.dropna(axis=0)

In [99]:
df.shape[0] * .7
# We keep 490 rows for training and remaining 230 for testing.

489.99999999999994

In [101]:
train, test = train_test_split(df, test_size = 0.3, random_state=1)

## Part 1.2.

In [110]:
stop_words = list(stopwords.words('turkish'))
stop_words

['acaba',
 'ama',
 'aslında',
 'az',
 'bazı',
 'belki',
 'biri',
 'birkaç',
 'birşey',
 'biz',
 'bu',
 'çok',
 'çünkü',
 'da',
 'daha',
 'de',
 'defa',
 'diye',
 'eğer',
 'en',
 'gibi',
 'hem',
 'hep',
 'hepsi',
 'her',
 'hiç',
 'için',
 'ile',
 'ise',
 'kez',
 'ki',
 'kim',
 'mı',
 'mu',
 'mü',
 'nasıl',
 'ne',
 'neden',
 'nerde',
 'nerede',
 'nereye',
 'niçin',
 'niye',
 'o',
 'sanki',
 'şey',
 'siz',
 'şu',
 'tüm',
 've',
 'veya',
 'ya',
 'yani']

In [112]:
vectorizer = TfidfVectorizer(stop_words=stop_words)
X = vectorizer.fit_transform(df.Text)

In [114]:
X_train, X_test, y_train, y_test = train_test_split(X, df.Classification, test_size=0.3, random_state=42)

## Part 1.3. Make SVM and test on C=0.001, 0.01, 0.1, 1, 10, 100, 1000

In [177]:
c_values = [0.001, 0.01, 0.1, 1, 10, 100, 1000]
for c in c_values:
    svm = SVC(kernel='linear', C=c)
    svm.fit(X_train, y_train)

    train_accuracy = svm.score(X_train, y_train)
    accuracy = svm.score(X_test, y_test)
    print(f"C = {c}\nAccuracy for: {accuracy:.2f}\nTrain Accuracy: {train_accuracy:.2f}")
    print(f"Number of support vectors for 2 classes: {svm.n_support_}\n")

C = 0.001
Accuracy for: 0.55
Train Accuracy: 0.61
Number of support vectors for 2 classes: [194 189]

C = 0.01
Accuracy for: 0.55
Train Accuracy: 0.61
Number of support vectors for 2 classes: [239 189]

C = 0.1
Accuracy for: 0.74
Train Accuracy: 0.78
Number of support vectors for 2 classes: [245 189]

C = 1
Accuracy for: 0.92
Train Accuracy: 0.99
Number of support vectors for 2 classes: [240 120]

C = 10
Accuracy for: 0.93
Train Accuracy: 1.00
Number of support vectors for 2 classes: [235 120]

C = 100
Accuracy for: 0.93
Train Accuracy: 1.00
Number of support vectors for 2 classes: [235 120]

C = 1000
Accuracy for: 0.93
Train Accuracy: 1.00
Number of support vectors for 2 classes: [235 120]



C = 0.001
Accuracy for: 0.55
Train Accuracy: 0.6142857142857143
Number of support vectors for 2 classes: [194 189]

C = 0.01
Accuracy for: 0.55
Train Accuracy: 0.6142857142857143
Number of support vectors for 2 classes: [239 189]

C = 0.1
Accuracy for: 0.74
Train Accuracy: 0.7795918367346939
Number of support vectors for 2 classes: [245 189]

C = 1
Accuracy for: 0.92
Train Accuracy: 0.9918367346938776
Number of support vectors for 2 classes: [240 120]

C = 10
Accuracy for: 0.93
Train Accuracy: 1.0
Number of support vectors for 2 classes: [235 120]

C = 100
Accuracy for: 0.93
Train Accuracy: 1.0
Number of support vectors for 2 classes: [235 120]

C = 1000
Accuracy for: 0.93
Train Accuracy: 1.0
Number of support vectors for 2 classes: [235 120]

## Part 1.4. Different Kernels (RBF, Polynomial)

In [179]:
# Kernels = poly’, ‘rbf
# POLY, deg = 1...3
# POLY coef0 = -1...1
svm = SVC(kernel='poly', C=1, degree=3, coef0=1)
svm.fit(X_train, y_train)

train_accuracy = svm.score(X_train, y_train)
accuracy = svm.score(X_test, y_test)
print(f"C = {c}\nAccuracy for: {accuracy:.2f}\nTrain Accuracy: {train_accuracy:.2f}")
print(f"Number of support vectors for 2 classes: {svm.n_support_}\n")

C = 1000
Accuracy for: 0.86
Train Accuracy: 1.00
Number of support vectors for 2 classes: [289 150]



In [197]:
param_grid = [
  {'degree': [1, 2, 3], 'coef0': [-1, 0, +1], 'kernel': ['poly']},
  {'gamma': [1, 1/2, 1/4], 'kernel': ['rbf']},
 ]

In [199]:
grid_svc = SVC()
clf = GridSearchCV(grid_svc, param_grid)
clf.fit(X_train, y_train)

GridSearchCV(estimator=SVC(),
             param_grid=[{'coef0': [-1, 0, 1], 'degree': [1, 2, 3],
                          'kernel': ['poly']},
                         {'gamma': [1, 0.5, 0.25], 'kernel': ['rbf']}])

In [200]:
clf.cv_results_
# mean_score_time

{'mean_fit_time': array([0.19445763, 0.14524951, 0.11818414, 0.1925343 , 0.1964663 ,
        0.19721055, 0.19181175, 0.19651108, 0.19832606, 0.19709172,
        0.19648056, 0.18502688]),
 'std_fit_time': array([0.00411593, 0.00419012, 0.0026259 , 0.00142559, 0.00321449,
        0.00280835, 0.00177508, 0.00347063, 0.00391681, 0.00327913,
        0.00303065, 0.00732516]),
 'mean_score_time': array([0.03916111, 0.03469191, 0.02717614, 0.03905201, 0.0456717 ,
        0.04678345, 0.03889184, 0.04053326, 0.04414372, 0.04214368,
        0.04068661, 0.04129553]),
 'std_score_time': array([0.00147129, 0.00056325, 0.00096088, 0.00138378, 0.00108942,
        0.00109869, 0.00142272, 0.0013501 , 0.00455526, 0.00108938,
        0.00080708, 0.00143554]),
 'param_coef0': masked_array(data=[-1, -1, -1, 0, 0, 0, 1, 1, 1, --, --, --],
              mask=[False, False, False, False, False, False, False, False,
                    False,  True,  True,  True],
        fill_value='?',
             dtype=obje

In [3]:
  # 'coef0': -1, 'degree': 1 = 0.92
  # 'coef0': -1, 'degree': 2 = 0.40
  # 'coef0': -1, 'degree': 3 = 0.90
  # 'coef0': 0, 'degree': 1 =  0.92
  # 'coef0': 0, 'degree': 2 =  0.88
  # 'coef0': 0, 'degree': 3 =  0.85
  # 'coef0': 1, 'degree': 1 =  0.92
  # 'coef0': 1, 'degree': 2 =  0.92
  # 'coef0': 1, 'degree': 3 =  0.91
  # 'gamma': 1 =               0.89
  # 'gamma': 0.5 =             0.89
  # 'gamma': 0.25 =            0.86